### ISSUE TITLE
Batching Request | Error Handling

### ISSUE THREAD

**Issue by thedamfr** [firebase/flashlight#74]
Hi !

Everything works well when the service locally, but in production, using `bonsai.io` I can't have concurrent writting.
So when I do batch update on my firebase database, I get the following : 
```
failed to delete firebase/user/<horrible-hash-here>: Error: Concurrent request limit exceeded. Please consider batching your requests, or contact support@bonsai.io for help.
````
For sure, more concurrent request means more billing. 
How would we be able to implement a throttling system or something similar ?
Maybe adding a retry queue for errors is an idea ?

**Comment by thedamfr** [comment]
Maybe we could job this with https://github.com/LewisJEllis/bee-queue ?

**Comment by katowulf** [comment]
I'm pretty sure the official ES node client already supports batching. Can you explain what you mean by "when I do batch update on my Firebase Database"? Some code/repro would be great here. If you already know the solution, PRs are also welcome.

**Comment by thedamfr** [comment]
Batch update on firebase : Upload a new json file that trigger hundreds of callbacks in flashlight.
I'm okay to PR, but I don't know which solution.

Here is a first idea : implement a in-memory queue with promises to make sure flashlight queue the callbacks action and only make one request to the ES at a time.

Another idea is to implement a Jobbing solution like beequeue. That would create a dependency but that would make flashlight really more consistent. We would be able to jobify every action in the callback, configure concurrency, handle errors. 

But that would create a dependency between flashlight and the jobbing lib. Maybe we can do a pluggable implementation.

**Comment by mbifulco** [comment]
Chiming in to say that this is an issue I'm running into also - I'm trying to index a firebase database that was preexisting and has hundreds of entries stored across different indices already - upon initial index, I'm greeted with loads of 

`Error: Concurrent request limit exceeded. Please consider batching your requests, or contact support@bonsai.io for help.` 

messages in my console.

**Comment by thedamfr** [comment]
hey @mbifulco quick and dirty fix is to to Pm2 launch a cluster of flashlight process. They will run several time. There will be no duplicates. Repeat process until it works :)

**Comment by mbifulco** [comment]
@thedamfr - interesting, I'll give it a shot.  What stops flashlight from wasting time re-indexing things that haven't changed in the database?

**Comment by thedamfr** [comment]
It will loose time. But you will be sure to be indexed at least once.

Le 29 sept. 2016 6:25 AM, "Mike Bifulco" <notifications@github.com> a
écrit :
>
> @thedamfr - interesting, I'll give it a shot. What stops flashlight from
wasting time re-indexing things that haven't changed in the database?

**Comment by mbifulco** [comment]
@katowulf any chance you could take a look at this again? It'd be really useful to have flashlight batch all of these index updates together.

I've got a firebase set up with a few thousand records in it that _very_ quickly overflows the concurrent request limit on bonsai.  The result is that every time I start the flashlight process, it hits the limit before it can push the most recent records into the es index.

I _believe_ this is because firebase's value/changed/added functions fire once for each child of my root nodes in the database.  If all of those could be batched together before sending them off to be indexed (especially when flashlight starts up), I'd avoid the concurrent request limit.

**Comment by katowulf** [comment]
So I'm pretty sure we've removed all the barriers to getting batching in place--we're now using the official ES client, which supports this, and capable of putting it in place. It will take some time and energy to decide how this should work (probably a debounce of some sort to batch them?) and to create a PR, although I think the code involved is pretty isolated: probably just [here](https://github.com/firebase/flashlight/blob/master/lib/PathMonitor.js#L38).

**Comment by mbifulco** [comment]
I agree.  I was just looking at using lodash's [throttle](https://lodash.com/docs/4.17.0#throttle) function to artificially slow things down.

**Comment by mbifulco** [comment]
Whoa - didn't know about `debounce`.  It makes way more sense.

**Comment by katowulf** [comment]
See also [Using refBuilder to improve indexing efficiency](https://github.com/firebase/flashlight#use-refbuilder-to-improve-indexing-efficiency), which is a decent workaround for big data sets until we do better batching.

**Comment by mkim871** [comment]
Any clear workaround explanation ? Im keep struggling with this problem

**Comment by katowulf** [comment]
Workaround explanation for what?

**Comment by leesolway** [comment]
Anyone figured a way around this issue as the code stands?

**Comment by katowulf** [comment]
Yes, using refBuilder as linked above is a "workaround" to better batching. The best workaround would, naturally, be a PR implementing batching.

**Comment by mike623** [comment]
For people who looking for solution

http://stackoverflow.com/questions/37456690/how-insure-indexing-every-object-with-firebase-flashlight-in-a-elasticsearch-bon

salute to [Anthony](https://stackoverflow.com/users/2733216/anthony)
